src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,514 @@
+        {
+            public int HashCode; // lower 31 bits of the hash code
+            public TKey Key;
+            public TValue Value;
+            public int Next; // the index of the next item in the same bucket
+        }
+
+        private int[] _buckets = Array.Empty<int>();
+        private Entry[] _entries = Array.Empty<Entry>();
+        private int _version;
+        private readonly IEqualityComparer<TKey> _comparer;
+        private KeyCollection _keys;
+        private ValueCollection _values;
+        
+        public int Count { get; private set; }
+
+        public IEqualityComparer<TKey> Comparer => _comparer ?? EqualityComparer<TKey>.Default;
+
+        public KeyCollection Keys => _keys ?? (_keys = new KeyCollection(this));
+
+        public ValueCollection Values => _values ?? (_values = new ValueCollection(this));
+
+        public TValue this[TKey key]
+        {
+            get => GetValue(key);
+            set => SetValue(key, value);
+        }
+
+        public TValue this[int index]
+        {
+            get => GetAt(index);
+            set => SetAt(index, value);
+        }
+
+        public OrderedDictionary()
+            : this(0, null)
+        {
+        }
+
+        public OrderedDictionary(int capacity)
+            : this(capacity, null)
+        {
+        }
+
+        public OrderedDictionary(IEqualityComparer<TKey> comparer)
+            : this(0, comparer)
+        {
+        }
+
+        public OrderedDictionary(int capacity, IEqualityComparer<TKey> comparer)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+            if (capacity > 0)
+            {
+                Resize(HashHelpers.GetPrime(capacity));
+            }
+
+            if (comparer != EqualityComparer<TKey>.Default)
+            {
+                _comparer = comparer;
+            }
+        }
+
+        public OrderedDictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection)
+            : this(collection, null)
+        {
+        }
+
+        public OrderedDictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEqualityComparer<TKey> comparer)
+            : this((collection as ICollection<KeyValuePair<TKey, TValue>>)?.Count ?? 0, comparer)
+        {
+            if (collection == null)
+            {
+                throw new ArgumentNullException(nameof(collection));
+            }
+
+            foreach (KeyValuePair<TKey, TValue> pair in collection)
+            {
+                Add(pair.Key, pair.Value);
+            }
+        }
+
+        public void Add(TKey key, TValue value) => TryInsert(null, key, value, InsertionBehavior.ThrowOnExisting);
+
+        public void Clear()
+        {
+            if (Count > 0)
+            {
+                Array.Clear(_buckets, 0, _buckets.Length);
+                Array.Clear(_entries, 0, Count);
+                Count = 0;
+            }
+            ++_version;
```

Comment:
I think `++_version` could be removed here if it's OK that enumeration just terminates. This was already done for Dictionary.